numba_dpex/experimental/kernel_dispatcher.py

```diff
@@ -0,0 +1,324 @@
+        try:
+            tp = typeof(val, Purpose.argument)
+            if isinstance(tp, types.Array) and not isinstance(tp, DpnpNdArray):
+                raise UnsupportedKernelArgumentError(
+                    type=str(type(val)), value=val
+                )
+        except ValueError:
+            tp = types.pyobject
+        else:
+            if tp is None:
+                tp = types.pyobject
+        self._types_active_call.append(tp)
+        return tp
+
+    def add_overload(self, cres, kernel_module):
+        args = tuple(cres.signature.args)
+        self.overloads[args] = kernel_module
+
+    def compile(self, sig) -> _KernelCompileResult:
+        disp = self._get_dispatcher_for_current_target()
+        if disp is not self:
+            return disp.compile(sig)
+
+        with ExitStack() as scope:
+            cres = None
+
+            def cb_compiler(dur):
+                if cres is not None:
+                    self._callback_add_compiler_timer(dur, cres)
+
+            def cb_llvm(dur):
+                if cres is not None:
+                    self._callback_add_llvm_timer(dur, cres)
+
+            scope.enter_context(
+                ev.install_timer("numba:compiler_lock", cb_compiler)
+            )
+            scope.enter_context(ev.install_timer("numba:llvm_lock", cb_llvm))
+            scope.enter_context(global_compiler_lock)
+
+            if not self._can_compile:
+                raise RuntimeError("compilation disabled")
+            # Use counter to track recursion compilation depth
+            with self._compiling_counter:
+                args, return_type = sigutils.normalize_signature(sig)
+                # Don't recompile if signature already exists
+                existing = self.overloads.get(tuple(args))
+                if existing is not None:
+                    return existing
+
+                # FIXME: Enable caching
+                # Add code to enable on disk caching of a binary spirv kernel
+                self._cache_misses[sig] += 1
+                ev_details = dict(
+                    dispatcher=self,
+                    args=args,
+                    return_type=return_type,
+                )
+                with ev.trigger_event("numba_dpex:compile", data=ev_details):
+                    try:
+                        kcres: _KernelCompileResult = self._compiler.compile(
+                            args, return_type
+                        )
+                    except errors.ForceLiteralArg as e:
+
+                        def folded(args, kws):
+                            return self._compiler.fold_argument_types(
+                                args, kws
+                            )[1]
+
+                        raise e.bind_fold_arguments(folded)
+                    self.add_overload(kcres.cres_or_error, kcres.kernel_module)
+
+                # FIXME: enable caching
+
+                return kcres.kernel_module
+
+    def __getitem__(self, args):
```

Comment:
Description mismatch. Accepts only one argument, and does not clones the objects. Also names are different.

Reply:
We do not need `__getitem__` and now it raised NotImplemented.